net/sched/p4tc/p4tc_cmds.c

```diff
@@ -1026,6 +1088,118 @@ static int validate_num_opnds(struct p4tc_cmd_operate *ope, u32 cmd_num_opnds)
 	return 0;
 }
 
+static struct p4tc_act_param *
+validate_act_param(struct p4tc_act *act, struct p4tc_cmd_operand *op,
+		   unsigned long *param_id, struct netlink_ext_ack *extack)
+{
+	struct p4tc_act_param *nparam;
+	struct p4tc_act_param *param;
+
+	param = idr_get_next_ul(&act->params_idr, param_id);
+	if (!param) {
+		NL_SET_ERR_MSG_MOD(extack,
+				   "Act has less runtime parameters than passed in call");
+		return ERR_PTR(-EINVAL);
+	}
+
+	nparam = kzalloc(sizeof(*nparam), GFP_KERNEL);
+	if (!nparam)
+		return ERR_PTR(-ENOMEM);
+
+	strscpy(nparam->name, param->name, ACTPARAMNAMSIZ);
+	nparam->id = *param_id;
+	nparam->value = op;
+	if (op->oper_datatype->typeid != param->type) {
+		NL_SET_ERR_MSG_MOD(extack, "Operand type differs from params");
+		return ERR_PTR(-EINVAL);
+	}
+	nparam->type = param->type;
+	nparam->flags |= P4TC_ACT_PARAM_FLAGS_ISDYN;
+
+	return nparam;
+}
+
+static int validate_act_params(struct net *net, struct p4tc_act *act,
+			       struct p4tc_cmd_operand *A,
+			       struct list_head *opnds_lst,
+			       struct list_head *params_lst,
+			       struct netlink_ext_ack *extack)
+{
+	struct p4tc_act_param *params[P4TC_MSGBATCH_SIZE] = {NULL};
+	unsigned long param_id = 0;
+	int i = 0;
+	struct p4tc_cmd_operand *kopnd;
+	int err;
+
+	kopnd = A;
+	list_for_each_entry_continue(kopnd, opnds_lst, oper_list_node) {
+		struct p4tc_act_param *nparam;
+
+		err = validate_operand(net, act, kopnd, extack);
+		if (err)
+			goto free_params;
+
+		nparam = validate_act_param(act, kopnd, &param_id, extack);
+		if (IS_ERR(nparam)) {
+			err = PTR_ERR(nparam);
+			goto free_params;
+		}
+
+		params[i] = nparam;
+		list_add_tail(&nparam->head, params_lst);
+		i++;
+		param_id++;
+	}
+
+	if (idr_get_next_ul(&act->params_idr, &param_id)) {
+		NL_SET_ERR_MSG_MOD(extack,
+				   "Act has more runtime params than passed in call");
+		return -EINVAL;
+	}
+
+	return 0;
+
+free_params:
+	while (i--)
+		kfree(params[i]);
+
+	return err;
+}
+
+static void free_intermediate_params_list(struct list_head *params_list)
+{
+	struct p4tc_act_param *nparam, *p;
+
+	list_for_each_entry_safe(nparam, p, params_list, head)
+		kfree(nparam);
+}
+
+/* Actions with runtime parameters don't have instance id. For example:
+ * cmd act ptables.myact.1 has instance id 1, while:
+ * cmd act ptables.myact param1 param2 ... doesn't specify instance id
```

Comment:
Almost sounds like you are saying cmd act ptables.myact.1 has instance id 1
is what you mean by runtime. Rewording:
-------
Actions with runtime parameters don't have instance ids (found in immedv2)
because the action is not created apriori. Example:
cmd act ptables.myact param1 param2 ... doesn't specify instance 
As noted, it is equivalent to treating an action like a function call with action attributes derived at runtime.
If these actions were already instantiated then immedv2 will have a non-zero value equal to the action index.
----

Reply:
Done here: https://github.com/p4tc-dev/linux-p4tc-pub/pull/45/files#diff-f70ae2b5731743e0d086cae79f9de3fd014258802d40560d8ca53b4ed1261ec3R1164